test/inductor/test_inductor_freezing_int4mm.py

```diff
@@ -0,0 +1,441 @@
+    -> split(int4_mm(x, cat(w1,w2,w3), gs, cat(sz1,sz2,sz3)), sizes)
+
+Requires: torchao, CUDA or MPS
+"""
+
+import unittest
+
+import torch
+from torch import nn
+from torch._dynamo.utils import counters
+from torch._inductor import config
+from torch._inductor.test_case import TestCase as InductorTestCase
+from torch._inductor.utils import run_and_get_code
+from torch.testing._internal.common_utils import skipIfRocm
+
+
+HAS_CUDA = torch.cuda.is_available()
+HAS_MPS = torch.backends.mps.is_available()
+HAS_CUDA_OR_MPS = HAS_CUDA or HAS_MPS
+DEVICE = "cuda" if HAS_CUDA else "mps" if HAS_MPS else "cpu"
+
+
+@unittest.skipIf(not HAS_CUDA_OR_MPS, "requires CUDA or MPS")
+@skipIfRocm
+class Int4MMFusionTests(InductorTestCase):
+    """Tests for INT4 GPU matmul fusion during freezing."""
+
+    device = DEVICE
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        # Import torchao here (after skip check) to avoid import errors on CPU CI
+        from torchao.quantization import Int4WeightOnlyConfig, quantize_
+
+        cls.Int4WeightOnlyConfig = Int4WeightOnlyConfig
+        cls.quantize_ = quantize_
+
+    def setUp(self):
+        super().setUp()
+        torch._dynamo.reset()
+        counters.clear()
+
+    def tearDown(self):
+        super().tearDown()
+        torch._dynamo.reset()
+
+    def test_int4_mm_fusion_three_way(self):
+        """Test 3-way INT4 matmul fusion (like Q/K/V projections)."""
+
+        class QKVProjection(nn.Module):
+            def __init__(self, dim=512):
+                super().__init__()
+                self.q_proj = nn.Linear(dim, dim, bias=False)
+                self.k_proj = nn.Linear(dim, dim, bias=False)
+                self.v_proj = nn.Linear(dim, dim, bias=False)
+
+            def forward(self, x):
+                return self.q_proj(x), self.k_proj(x), self.v_proj(x)
+
+        model = QKVProjection(dim=512).eval().to(self.device).to(torch.bfloat16)
+
+        # Quantize to INT4
+        int4_config = self.Int4WeightOnlyConfig(
+            group_size=128,
+            int4_packing_format="tile_packed_to_4d",
+        )
+        self.quantize_(model, int4_config)
+
+        x = torch.randn(2, 128, 512, device=self.device, dtype=torch.bfloat16)
+
+        # Get reference output
+        with torch.no_grad():
+            ref = model(x)
+
+        # Compile with freezing
+        with config.patch(
+            {
+                "freezing": True,
+                "freezing_discard_parameters": False,
+                "implicit_fallbacks": True,
+            }
+        ):
+
+            @torch.compile()
+            def foo(mod, inp):
+                return mod(inp)
+
+            with torch.no_grad():
+                out, code = run_and_get_code(foo, model, x)
+
+        # Verify output correctness
+        for i in range(3):
+            self.assertEqual(ref[i], out[i])
+
+        # Count INT4 matmul calls - should be 1 (fused) not 3
```

Comment:
There is an existing utility function, `FileCheck().check_cout`, for this kind of checking, see https://github.com/pytorch/pytorch/blob/c3c11b794173a4e6641f27ead4ce20ea203a7ac5/test/inductor/test_torchinductor.py#L11964